docs/settings.rst
.. _settings:

Settings
########

Customize your provider so fit your project needs.

SITE_URL
========

REQUIRED. The OP server url.

``str``. For example ``http://localhost:8000``.

LOGIN_URL
=========

REQUIRED. Used to log the user in. `Read more in Django docs <https://docs.djangoproject.com/en/1.7/ref/settings/#login-url>`_

``str``. Default is ``/accounts/login/``.

OIDC_AFTER_USERLOGIN_HOOK
=========================

OPTIONAL. A string with the location of your function. Provide a way to plug into the process after the user has logged in, typically to perform some business logic.

Default is::

    def default_hook_func(request, user, client):
        return None

Return ``None`` if you want to continue with the flow.

The typical situation will be checking some state of the user or maybe redirect him somewhere.
With request you have access to all OIDC parameters. Remember that if you redirect the user to another place then you need to take him back to the authorize endpoint (use ``request.get_full_path()`` as the value for a "next" parameter).

OIDC_CODE_EXPIRE
================

OPTIONAL. Code object expiration after been delivered.

``int``. Expressed in seconds. Default is ``60*10``.

OIDC_EXTRA_SCOPE_CLAIMS
=======================

OPTIONAL. A string with the location of your class. Default is ``oidc_provider.lib.claims.AbstractScopeClaims``.

Used to add extra scopes specific for your app. This class MUST inherit ``AbstractScopeClaims``.

OpenID Connect Clients will use scope values to specify what access privileges are being requested for Access Tokens.

`Here <http://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims>`_ you have the standard scopes defined by the protocol.

Check out an example of how to implement it::

    from oidc_provider.lib.claims import AbstractScopeClaims

    class MyAppScopeClaims(AbstractScopeClaims):

        def setup(self):
            # Here you can load models that will be used
            # in more than one scope for example.
            # print self.user
            # print self.scopes
            try:
                self.some_model = SomeModel.objects.get(user=self.user)
            except SomeModel.DoesNotExist:
                # Create an empty model object.
                self.some_model = SomeModel()

        def scope_books(self, user):

            # Here you can search books for this user.

            dic = {
                'books_readed': books_readed_count,
            }

            return dic

You can create our own scopes using the convention:

``def scope_SCOPENAMEHERE(self, user):``

If a field is empty or ``None`` will be cleaned from the response.

OIDC_IDTOKEN_EXPIRE
===================

OPTIONAL. Token object expiration after been delivered.

``int``. Expressed in seconds. Default is ``60*10``.

OIDC_IDTOKEN_SUB_GENERATOR
==========================

OPTIONAL. A string with the location of your function. ``sub`` is a locally unique and never reassigned identifier within the Issuer for the End-User, which is intended to be consumed by the Client.

The function receives a ``user`` object and returns a unique ``string`` for the given user.

Default is::

    def default_sub_generator(user):

        return str(user.id)

OIDC_SKIP_CONSENT_ALWAYS
========================

OPTIONAL. If enabled, the Server will NEVER ask the user for consent.

``bool``. Default is ``False``.

OIDC_SKIP_CONSENT_ENABLE
========================

OPTIONAL. If enabled, the Server will save the user consent given to a specific client, so that user won't be prompted for the same authorization multiple times.

``bool``. Default is ``True``.

OIDC_SKIP_CONSENT_EXPIRE
========================

OPTIONAL. User consent expiration after been granted.

``int``. Expressed in days. Default is ``30*3``.

OIDC_TOKEN_EXPIRE
=================

OPTIONAL. Token object expiration after been created.

``int``. Expressed in seconds. Default is ``60*60``.

OIDC_USERINFO
=============

OPTIONAL. A string with the location of your class. Read **Standard Claims** section.


OIDC_ID_TOKEN_PROCESSING_HOOK
=============================

OPTIONAL. A string with the location of your hook.
Used to add extra dictionary values specific for your app into id_token.




docs/sections/settings.rst
.. _settings:

Settings
########

Customize your provider so fit your project needs.

SITE_URL
========

REQUIRED. ``str``. The OP server url.

For example ``http://localhost:8000``.

LOGIN_URL
=========

REQUIRED. ``str``. Used to log the user in. `Read more in Django docs <https://docs.djangoproject.com/en/1.7/ref/settings/#login-url>`_

``str``. Default is ``/accounts/login/``.

OIDC_AFTER_USERLOGIN_HOOK
=========================

OPTIONAL. ``str``. A string with the location of your function. Provide a way to plug into the process after the user has logged in, typically to perform some business logic.

Default is::

    def default_hook_func(request, user, client):
        return None

Return ``None`` if you want to continue with the flow.

The typical situation will be checking some state of the user or maybe redirect him somewhere.
With request you have access to all OIDC parameters. Remember that if you redirect the user to another place then you need to take him back to the authorize endpoint (use ``request.get_full_path()`` as the value for a "next" parameter).

OIDC_CODE_EXPIRE
================

OPTIONAL. ``int``. Code object expiration after been delivered.

Expressed in seconds. Default is ``60*10``.

OIDC_EXTRA_SCOPE_CLAIMS
=======================

OPTIONAL. ``str``. A string with the location of your class. Default is ``oidc_provider.lib.claims.AbstractScopeClaims``.

Used to add extra scopes specific for your app. This class MUST inherit ``AbstractScopeClaims``.

OpenID Connect Clients will use scope values to specify what access privileges are being requested for Access Tokens.

`Here <http://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims>`_ you have the standard scopes defined by the protocol.

Check out an example of how to implement it::

    from oidc_provider.lib.claims import AbstractScopeClaims

    class MyAppScopeClaims(AbstractScopeClaims):

        def setup(self):
            # Here you can load models that will be used
            # in more than one scope for example.
            # print self.user
            # print self.scopes
            try:
                self.some_model = SomeModel.objects.get(user=self.user)
            except SomeModel.DoesNotExist:
                # Create an empty model object.
                self.some_model = SomeModel()

        def scope_books(self, user):

            # Here you can search books for this user.

            dic = {
                'books_readed': books_readed_count,
            }

            return dic

You can create our own scopes using the convention:

``def scope_SCOPENAMEHERE(self, user):``

If a field is empty or ``None`` will be cleaned from the response.

OIDC_IDTOKEN_EXPIRE
===================

OPTIONAL. ``int``. Token object expiration after been delivered.

Expressed in seconds. Default is ``60*10``.

OIDC_IDTOKEN_PROCESSING_HOOK
============================

OPTIONAL. ``str``. A string with the location of your function hook.
Here you can add extra dictionary values specific for your app into id_token.

The function receives a ``id_token`` dictionary and ``user`` instance 
and returns it with additional fields.

Default is::

    def default_idtoken_processing_hook(id_token, user):

        return return id_token

OIDC_IDTOKEN_SUB_GENERATOR
==========================

OPTIONAL. ``str``. A string with the location of your function. ``sub`` is a locally unique and never reassigned identifier within the Issuer for the End-User, which is intended to be consumed by the Client.

The function receives a ``user`` object and returns a unique ``string`` for the given user.

Default is::

    def default_sub_generator(user):

        return str(user.id)

OIDC_SKIP_CONSENT_ALWAYS
========================

OPTIONAL. ``bool``. If enabled, the Server will NEVER ask the user for consent.

Default is ``False``.

OIDC_SKIP_CONSENT_ENABLE
========================

OPTIONAL. ``bool``. If enabled, the Server will save the user consent given to a specific client, so that user won't be prompted for the same authorization multiple times.

Default is ``True``.

OIDC_SKIP_CONSENT_EXPIRE
========================

OPTIONAL. ``int``. User consent expiration after been granted.

Expressed in days. Default is ``30*3``.

OIDC_TOKEN_EXPIRE
=================

OPTIONAL. ``int``. Token object expiration after been created.

Expressed in seconds. Default is ``60*60``.

OIDC_USERINFO
=============

OPTIONAL. ``str``. A string with the location of your class. Read **Standard Claims** section.


CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.

### [Unreleased]

##### Added
- Support OAuth2 requests.
- Decorator for protecting views with OAuth2.
- Setting OIDC_IDTOKEN_PROCESSING_HOOK.

### [0.2.5] - 2016-02-03

##### Added
- Setting OIDC_SKIP_CONSENT_ALWAYS.

##### Changed
- Removing OIDC_RSA_KEY_FOLDER setting. Moving RSA Keys to the database.
- Update pyjwkest to version 1.1.0.

##### Fixed
- Nonce parameter missing on the decide form.
- Set Allow-Origin header to jwks endpoint.

### [0.2.4] - 2016-01-20

##### Added
- Auto-generation of client ID and SECRET using the admin.
- Validate nonce parameter when using Implicit Flow.

##### Fixed
- Fixed generating RSA key by ignoring value of OIDC_RSA_KEY_FOLDER.
- Make OIDC_AFTER_USERLOGIN_HOOK and OIDC_IDTOKEN_SUB_GENERATOR to be lazy imported by the location of the function.
- Problem with a function that generate urls for the /.well-known/openid-configuration/ endpoint.

### [0.2.3] - 2016-01-06

##### Added
- Make user and client unique on UserConsent model.
- Support for URL's without end slash.

##### Changed
- Upgrade pyjwkest to version 1.0.8.

##### Fixed
- String format error in models.
- Redirect to non http urls fail (for Mobile Apps).

### [0.2.1] - 2015-10-21

##### Added
- Refresh token flow.

##### Changed
- Upgrade pyjwkest to version >= 1.0.6.

##### Fixed
- Unicode error in Client model.
- Bug in creatersakey command (when using Python 3).
- Bug when updating pyjwkest version.

### [0.2.0] - 2015-09-25

##### Changed
- UserInfo model was removed. Now you can add your own model using OIDC_USERINFO setting.

##### Fixed
- ID token does NOT contain kid.

### [0.1.2] - 2015-08-04

##### Added
- Add token_endpoint_auth_methods_supported to discovery.

##### Fixed
- Missing commands folder in setup file.

### [0.1.1] - 2015-07-31

##### Added
- Sending access_token as query string parameter in UserInfo Endpoint.
- Support HTTP Basic client authentication.

##### Changed
- Use models setting instead of User.

##### Fixed
- In python 2: "aud" and "nonce" parameters didn't appear in id_token.

### [0.1.0] - 2015-07-17

##### Added
- Now id tokens are signed/encrypted with RS256.
- Command for easily generate random RSA key.
- Jwks uri to discovery endpoint.
- id_token_signing_alg_values_supported to discovery endpoint.

##### Fixed
- Nonce support for both Code and Implicit flow.

### [0.0.7] - 2015-07-06

##### Added
- Support for Python 3.
- Way of remember user consent and skipt it (OIDC_SKIP_CONSENT_ENABLE).
- Setting OIDC_SKIP_CONSENT_EXPIRE.

##### Changed
- Now OIDC_EXTRA_SCOPE_CLAIMS must be a string, to be lazy imported.

### [0.0.6] - 2015-06-16

##### Added
- Better naming for models in the admin.

##### Changed
- Now tests run without the need of a project configured.

##### Fixed
- Error when returning address_formatted claim.

### [0.0.5] - 2015-05-09

##### Added
- Support for Django 1.8.

##### Fixed
- Validation of scope in UserInfo endpoint.

### [0.0.4] - 2015-04-22

##### Added
- Initial migrations.

##### Fixed
- Important bug with id_token when using implicit flow.
- Validate Code expiration in Auth Code Flow.
- Validate Access Token expiration in UserInfo endpoint.

### [0.0.3] - 2015-04-15

##### Added
- Normalize gender field in UserInfo.

##### Changed
- Make address_formatted a property inside UserInfo.

##### Fixed
- Important bug in claims response.

### [0.0.2] - 2015-03-26

##### Added
- Setting OIDC_AFTER_USERLOGIN_HOOK.

##### Fixed
- Tests failing because an incorrect tag in one template.

### [0.0.1] - 2015-03-13

##### Added
- Provider Configuration Information endpoint.
- Setting OIDC_IDTOKEN_SUB_GENERATOR.

##### Changed
- Now use setup in OIDC_EXTRA_SCOPE_CLAIMS setting.

### [0.0.0] - 2015-02-26


oidc_provider/settings.py
import importlib

from django.conf import settings


class DefaultSettings(object):

    @property
    def LOGIN_URL(self):
        """
        REQUIRED. Used to log the user in.
        """
        return None

    @property
    def SITE_URL(self):
        """
        REQUIRED. The OP server url.
        """
        return None

    @property
    def OIDC_AFTER_USERLOGIN_HOOK(self):
        """
        OPTIONAL.  Provide a way to plug into the process after
        the user has logged in, typically to perform some business logic.
        """
        return 'oidc_provider.lib.utils.common.default_after_userlogin_hook'

    @property
    def OIDC_CODE_EXPIRE(self):
        """
        OPTIONAL. Code expiration time expressed in seconds.
        """
        return 60*10

    @property
    def OIDC_EXTRA_SCOPE_CLAIMS(self):
        """
        OPTIONAL. A string with the location of your class.
        Used to add extra scopes specific for your app. 
        """
        return 'oidc_provider.lib.claims.AbstractScopeClaims'

    @property
    def OIDC_IDTOKEN_EXPIRE(self):
        """
        OPTIONAL. Id token expiration time expressed in seconds.
        """
        return 60*10

    @property
    def OIDC_IDTOKEN_SUB_GENERATOR(self):
        """
        OPTIONAL. Subject Identifier. A locally unique and never
        reassigned identifier within the Issuer for the End-User,
        which is intended to be consumed by the Client.
        """
        return 'oidc_provider.lib.utils.common.default_sub_generator'

    @property
    def OIDC_SKIP_CONSENT_ALWAYS(self):
        """
        OPTIONAL. If enabled, the Server will NEVER ask the user for consent.
        """
        return False

    @property
    def OIDC_SKIP_CONSENT_ENABLE(self):
        """
        OPTIONAL. If enabled, the Server will save the user consent
        given to a specific client, so that user won't be prompted for
        the same authorization multiple times.
        """
        return True

    @property
    def OIDC_SKIP_CONSENT_EXPIRE(self):
        """
        OPTIONAL. User consent expiration after been granted.
        """
        return 30*3

    @property
    def OIDC_TOKEN_EXPIRE(self):
        """
        OPTIONAL. Token object expiration after been created.
        Expressed in seconds.
        """
        return 60*60

    @property
    def OIDC_USERINFO(self):
        """
        OPTIONAL. A string with the location of your class.
        Used to add extra scopes specific for your app. 
        """
        return 'oidc_provider.lib.utils.common.DefaultUserInfo'

    @property
    def OIDC_IDTOKEN_PROCESSING_HOOK(self):
        """
        OPTIONAL. A string with the location of your hook.
        Used to add extra dictionary values specific for your app into id_token.
        """
        return 'oidc_provider.lib.utils.common.default_idtoken_processing_hook'

default_settings = DefaultSettings()


def import_from_str(value):
    """
    Attempt to import a class from a string representation.
    """
    try:
        parts = value.split('.')
        module_path, class_name = '.'.join(parts[:-1]), parts[-1]
        module = importlib.import_module(module_path)
        return getattr(module, class_name)
    except ImportError as e:
        msg = 'Could not import %s for settings. %s: %s.' % (value, e.__class__.__name__, e)
        raise ImportError(msg)


def get(name, import_str=False):
    """
    Helper function to use inside the package.
    """
    value = None
    try:
        value = getattr(default_settings, name)
        value = getattr(settings, name)
    except AttributeError:
        if value is None:
            raise Exception('You must set ' + name + ' in your settings.')

    value = import_from_str(value) if import_str else value

    return value
